### DISCUSSION TITLE
How to remove all spin matching processing in Pyscf UHF calculation?

### DISCUSSION THREAD

**Discussion by zhaoyili401** [pyscf/pyscf#2626] [Q&A] [answered]
In the actual calculation, we found that pyscf UHF calculation seems to default to a certain spin matching processing, I want to completely turn off to achieve a complete UHF calculation, how can I do it

**Comment by jeanwsr** [comment]
There's no such spin matching or restriction in pyscf, so nothing to turn off. You may found that a UHF calculation for singlet state has no spin polarization, which is normal when you didn't apply a spin polarized (aka symmetry broken) initial guess.

**Comment by zhaoyili401** [comment]
> There's no such spin matching or restriction in pyscf, so nothing to turn off. You may found that a UHF calculation for singlet state has no spin polarization, which is normal when you didn't apply a spin polarized (aka symmetry broken) initial guess.

Dear respondent, hello. I calculated a double state molecule with a single electron, and the result obtained from UHF calculation using pyscf is different from the energy calculated using CFOUR. Moreover, the result obtained from CFOUR has more spin contamination, while the result obtained from pyscf has less spin contamination but slightly higher energy. We speculate that there may be some spin matching processing in the UHF calculation of pyscf, and we want to remove it and obtain the same result as CFOUR。

**Comment by jeanwsr** [comment, marked as answer]
There's no such spin matching. 

Converging to different solutions of UHF with different program/algorithms is also normal. You may check the UHF wavefunction stability first.

**Comment by zhaoyili401** [comment]
> There's no such spin matching.
> 
> Converging to different solutions of UHF with different program/algorithms is also normal. You may check the UHF wavefunction stability first.

OK,thank you,I get the stable wave function.This is what I want

**Comment by 1234zou** [comment]
Hi, your original problem has been solved. Here I want to show that there is no need to perform two independent calculations both in PySCF and CFOUR. If you have accomplished the CFOUR job, you can transfer CFOUR molecular orbitals (MOs) to PySCF, so that UHF in PySCF will be converged to exactly the same UHF solution. CFOUR usually generates a MOLDEN file after the job is finished, then you can run the following commands
```
mv MOLDEN O2_singlet.molden
molden2fch O2_singlet.molden -cfour
```
to generate O2_singlet.fch file, which is a Gaussian .fch(k) format file (but no Gaussian software is required). Here a singlet O2 is used as an example. [molden2fch](https://jeanwsr.gitlab.io/mokit-doc-mdbook/chap4-5.html#4537-molden2fch) is a utility from the open-source package [MOKIT](https://gitlab.com/jxzou/mokit), and it can be used to transferred MOs among various quantum chemistry packages (including CFOUR-v2.1). Now you can write a Python script to read MOs (in fact, all information) from this .fch file
```
from pyscf import gto, scf, lib
from mokit.lib.gaussian import load_mol_from_fch, mo_fch2py

lib.num_threads(2) # 2 CPU cores
fchname = 'O2_singlet.fch'
mol = load_mol_from_fch(fchname)

mf = scf.UHF(mol)
mf.max_cycle = 1
mf.kernel()

mf.mo_coeff = mo_fch2py(fchname)
dm = mf.make_rdm1()
mf.max_cycle = 10
mf.kernel(dm0=dm)
```

Submit the PySCF job
```
python O2_singlet.py >O2_singlet.out 2>&1
```

And you will see the UHF in PySCF is converged immediately, with negligible energy change
```
init E= -149.651906484745
  alpha nocc = 8  HOMO = -0.505823828471482  LUMO = 0.104523375811265
  beta  nocc = 8  HOMO = -0.505823828476212  LUMO = 0.104523375815523
cycle= 1 E= -149.651906476793  delta_E= 7.95e-09  |g|= 8.54e-09  |ddm|= 1.86e-08
  alpha nocc = 8  HOMO = -0.505823822700943  LUMO = 0.104523382217903
  beta  nocc = 8  HOMO = -0.505823822702403  LUMO = 0.10452338221891
cycle= 2 E= -149.651906476793  delta_E= 5.68e-14  |g|= 2.94e-09  |ddm|= 4.42e-09
  alpha nocc = 8  HOMO = -0.505823824952158  LUMO = 0.10452338033396
  beta  nocc = 8  HOMO = -0.505823824952622  LUMO = 0.104523380334194
Extra cycle  E= -149.651906476793  delta_E= -8.53e-14  |g|= 1.02e-09  |ddm|= 2.17e-09
converged SCF energy = -149.651906476793  <S^2> = 1.0231814  2S+1 = 2.2567068
```

This does not mean that CFOUR UHF is better than PySCF, but just saving the computational time and saving time for checking/comparing results of two packages. Also, one can check the CFOUR UHF wave function stability using PySCF.

**Comment by zhaoyili401** [comment]
> Hi, your original problem has been solved. Here I want to show that there is no need to perform two independent calculations both in PySCF and CFOUR. If you have accomplished the CFOUR job, you can transfer CFOUR molecular orbitals (MOs) to PySCF, so that UHF in PySCF will be converged to exactly the same UHF solution. CFOUR usually generates a MOLDEN file after the job is finished, then you can run the following commands
> 
> ```
> mv MOLDEN O2_singlet.molden
> molden2fch O2_singlet.molden -cfour
> ```
> 
> to generate O2_singlet.fch file, which is a Gaussian .fch(k) format file (but no Gaussian software is required). Here a singlet O2 is used as an example. [molden2fch](https://jeanwsr.gitlab.io/mokit-doc-mdbook/chap4-5.html#4537-molden2fch) is a utility from the open-source package [MOKIT](https://gitlab.com/jxzou/mokit), and it can be used to transferred MOs among various quantum chemistry packages (including CFOUR-v2.1). Now you can write a Python script to read MOs (in fact, all information) from this .fch file
> 
> ```
> from pyscf import gto, scf, lib
> from mokit.lib.gaussian import load_mol_from_fch, mo_fch2py
> 
> lib.num_threads(2) # 2 CPU cores
> fchname = 'O2_singlet.fch'
> mol = load_mol_from_fch(fchname)
> 
> mf = scf.UHF(mol)
> mf.max_cycle = 1
> mf.kernel()
> 
> mf.mo_coeff = mo_fch2py(fchname)
> dm = mf.make_rdm1()
> mf.max_cycle = 10
> mf.kernel(dm0=dm)
> ```
> 
> Submit the PySCF job
> 
> ```
> python O2_singlet.py >O2_singlet.out 2>&1
> ```
> 
> And you will see the UHF in PySCF is converged immediately, with negligible energy change
> 
> ```
> init E= -149.651906484745
>   alpha nocc = 8  HOMO = -0.505823828471482  LUMO = 0.104523375811265
>   beta  nocc = 8  HOMO = -0.505823828476212  LUMO = 0.104523375815523
> cycle= 1 E= -149.651906476793  delta_E= 7.95e-09  |g|= 8.54e-09  |ddm|= 1.86e-08
>   alpha nocc = 8  HOMO = -0.505823822700943  LUMO = 0.104523382217903
>   beta  nocc = 8  HOMO = -0.505823822702403  LUMO = 0.10452338221891
> cycle= 2 E= -149.651906476793  delta_E= 5.68e-14  |g|= 2.94e-09  |ddm|= 4.42e-09
>   alpha nocc = 8  HOMO = -0.505823824952158  LUMO = 0.10452338033396
>   beta  nocc = 8  HOMO = -0.505823824952622  LUMO = 0.104523380334194
> Extra cycle  E= -149.651906476793  delta_E= -8.53e-14  |g|= 1.02e-09  |ddm|= 2.17e-09
> converged SCF energy = -149.651906476793  <S^2> = 1.0231814  2S+1 = 2.2567068
> ```
> 
> This does not mean that CFOUR UHF is better than PySCF, but just saving the computational time and saving time for checking/comparing results of two packages. Also, one can check the CFOUR UHF wave function stability using PySCF.

Thank you！This is a very good and time saving method.In your python code your logic is to use CFOUR MODLEN file for molecular structure construction, then perform a UHF calculation, and then take the coefficient matrix in the MODLEN file and the first order density matrix in the first UHF calculation to recalculate? I would like to ask what step mf.mo_coeff = mo_fch2py(fchname) is. In my calculation, CFOUR has the lowest ground state energy, but the spin pollution is more serious. Before seeing your answer, I used the loop optimization function in pyscf stability analysis to get the wave function I wanted, but I found that there was no way to remove the optimized wave function chk file or diis file to restart the calculation, do you know how to do it?

**Comment by 1234zou** [comment]
`mol = load_mol_from_fch(fchname)` means creating the `mol` object directly from information in the specified .fch(k) file. One does not need to manually write the definition of `mol` (elements, coordinates, basis set, etc).

`mf.mo_coeff = mo_fch2py(fchname)` means reading MO coefficients from the specified .fch(k) file and convert them into PySCF AO (atomic orbital) conventions. `mo_coeff` is an attribute of `mf`, which means MO coefficients.

There is no 'spin pollution', but 'spin contamination'. I know these two words are similar when using Chinese, but the terminology 'spin contamination' is unique.

It seems that your last sentence 'get the wave function I wanted' and 'there was no way to remove the optimized wave function chk file or diis file to restart the calculation' are contradictory. I do not know what does it mean. You are supposed to show sufficient details for a new question.

**Comment by zhaoyili401** [comment]
> `mol = load_mol_from_fch(fchname)` means creating the `mol` object directly from information in the specified .fch(k) file. One does not need to manually write the definition of `mol` (elements, coordinates, basis set, etc).
> 
> `mf.mo_coeff = mo_fch2py(fchname)` means reading MO coefficients from the specified .fch(k) file and convert them into PySCF AO (atomic orbital) conventions. `mo_coeff` is an attribute of `mf`, which means MO coefficients.
> 
> There is no 'spin pollution', but 'spin contamination'. I know these two words are similar when using Chinese, but the terminology 'spin contamination' is unique.
> 
> It seems that your last sentence 'get the wave function I wanted' and 'there was no way to remove the optimized wave function chk file or diis file to restart the calculation' are contradictory. I do not know what does it mean. You are supposed to show sufficient details for a new question.

Thanks for your correction. The new problem I want to express is that I used the stability loop orbits optimization function in [pyscf/examples/scf/17-stability.py](https://github.com/pyscf/pyscf/blob/master/examples/scf/17-stability.py). I want to get the optimized .chk and diis.h5 files so that I can quickly get the optimized wave function I want again. But I don't know how to write.I tried to write after optimizing mf,
```
mf = stable_opt_internal(mf)
mf.chkfile='1.chk'
```
but I did not get the.chk file.

**Comment by jeanwsr** [comment]
`mf.chkfile='1.chk'` is only a change to attribute, not an operation. To perform a single operation for writing wfn, you may call the `scf.chkfile.dump_scf` function. Another way is to change the attribute before the `mf.kernel` called in your script.

**Comment by zhaoyili401** [comment]
> `mf.chkfile='1.chk'` is only a change to attribute, not an operation. To perform a single operation for writing wfn, you may call the `scf.chkfile.dump_scf` function. Another way is to change the attribute before the `mf.kernel` called in your script.

It's okay to change the properties before mf.kernel (), but the stability optimization will re-iterate the SCF iteration, and I want the stability optimized chk file

**Comment by jeanwsr** [comment]
You can re-write the `stable_opt_internal` function to assign the attribute inside the loop, or wherever you want.

**Comment by zhaoyili401** [comment]
> You can re-write the `stable_opt_internal` function to assign the attribute inside the loop, or wherever you want.

OK Thank you！
